### PR TITLE
Split at stop points

### PR DESCRIPTION
**Pull request by Lewis-Seiden**

### CODE REVIEW

**Comment by shueja** [comment]
A thought:
Do we want all stop points to necessarily be split points?

**Comment by Lewis-Seiden** [comment]
> A thought: Do we want all stop points to necessarily be split points?

i think so? a user can manually add split segments together again afterwards anyways

**Review by shueja — CHANGES_REQUESTED** [review]
Still missing c++ equivalents on the group loading

File: choreolib/src/main/native/cpp/choreo/lib/Choreo.cpp [hunk 1]
```diff
@@ -33,6 +34,29 @@ ChoreoTrajectory Choreo::GetTrajectory(std::string_view trajName) {
   return traj;
 }
 
+ChoreoTrajectory[] Choreo::GetTrajectoryGroup(std::string_view trajName) {
+  const std::filesystem::path trajDir{
+      fmt::format("{}/choreo", frc::filesystem::GetDeployDirectory())};
+  int segmentCount = 0;
+  for (auto const& dir_entry : std::filesystem::directory_iterator{sandbox}) {
```
Comment (calcmogul) [inline choreolib/src/main/native/cpp/choreo/lib/Choreo.cpp:41]:
```suggestion
  for (const auto& dir_entry : std::filesystem::directory_iterator{sandbox}) {
```

File: choreolib/src/main/native/cpp/choreo/lib/Choreo.cpp [hunk 2]
```diff
@@ -33,6 +34,29 @@ ChoreoTrajectory Choreo::GetTrajectory(std::string_view trajName) {
...
+    if (dir_entry.is_regular_file() &&
+        std::regex::regex_match(dir_entry.path().stem().string(),
+                                trajName + "\\.\\d+\\.traj")) {
+      segmentCount = 0;
```
Comment (calcmogul) [inline choreolib/src/main/native/cpp/choreo/lib/Choreo.cpp:45]: This look is a no-op.

File: choreolib/src/main/native/include/choreo/lib/Choreo.h [hunk 3]
```diff
@@ -34,6 +34,22 @@ class Choreo {
...
+   * .traj files to be placed in
+   * src/main/deploy/choreo/[trajName].[segmentNumber].traj . <br/> This method
+   * determines the number of parts to load by counting the files that match the
+   * pattern "trajName.X.traj", where X is a string of digits. Let this count be
+   * N. It then attempts to load "trajName.1.traj" through "trajName.N.traj",
+   * consecutively counting up. If any of these files cannot be loaded, the
+   * method throws a runtime error.
+   * @param trajName the path name in Choreo. Do not include ".traj" here.
+   * @return The array of segments, in order. Will throw runtime error if any
+   * segment does not exist.
```
Comment (calcmogul) [inline choreolib/src/main/native/include/choreo/lib/Choreo.h:48]:
```suggestion
   * Loads the split parts of the specified trajectory.
   *
   * ChoreoLib expects split .traj files to be placed in
   * src/main/deploy/choreo/[trajName].[segmentNumber].traj.
   *
   * This method determines the number of parts to load by counting the files
   * that match the pattern "trajName.X.traj", where X is a string of digits.
   * Let this count be N. It then attempts to load "trajName.1.traj" through
   * "trajName.N.traj", consecutively counting up.
   *
   * @param trajName the path name in Choreo. Do not include ".traj" here.
   * @return The array of segments, in order.
   * @throws std::runtime_error If any files cannot be loaded.
```

**Review by calcmogul — CHANGES_REQUESTED** [review]

**Review by calcmogul — CHANGES_REQUESTED** [review]

File: choreolib/src/main/native/cpp/choreo/lib/Choreo.cpp [hunk 4]
```diff
@@ -33,6 +34,29 @@ ChoreoTrajectory Choreo::GetTrajectory(std::string_view trajName) {
...
+  int segmentCount = 0;
+  for (const auto& dir_entry : std::filesystem::directory_iterator{sandbox}) {
+    if (dir_entry.is_regular_file() &&
+        std::regex::regex_match(dir_entry.path().stem().string(),
+                                trajName + "\\.\\d+\\.traj")) {
+      segmentCount = 0;
+    }
+  }
+  ChoreoTrajectory group[] = ChoreoTrajectory[segmentCount];
+  for (int i = 1; i <= segmentCount; i++) {
```
Comment (calcmogul) [inline choreolib/src/main/native/cpp/choreo/lib/Choreo.cpp:49]:
```suggestion
  for (int i = 1; i <= segmentCount; ++i) {
```

File: choreolib/src/main/java/com/choreo/lib/Choreo.java [hunk 5]
```diff
@@ -38,6 +39,36 @@ public static ChoreoTrajectory getTrajectory(String trajName) {
...
+   * @param trajName The path name in Choreo for this trajectory.
+   * @return The ArrayList of segments, in order, or null.
+   */
+  public static ArrayList<ChoreoTrajectory> getTrajectoryGroup(String trajName) {
+    var traj_dir = new File(Filesystem.getDeployDirectory(), "choreo");
+    File[] files =
+        traj_dir.listFiles((file) -> file.getName().matches(trajName + "\\.\\d+\\.traj"));
+    int segmentCount = files.length;
+    var trajs = new ArrayList<ChoreoTrajectory>();
+    for (int i = 1; i <= segmentCount; i++) {
```
Comment (calcmogul) [inline choreolib/src/main/java/com/choreo/lib/Choreo.java:59]:
```suggestion
    for (int i = 1; i <= segmentCount; ++i) {
```

File: choreolib/src/main/java/com/choreo/lib/Choreo.java [hunk 6]
```diff
@@ -38,6 +39,36 @@ public static ChoreoTrajectory getTrajectory(String trajName) {
...
+   * This method determines the number of parts to load by counting the files that match the pattern
+   * "trajName.X.traj", where X is a string of digits. Let this count be N. It then attempts to load
+   * "trajName.1.traj" through "trajName.N.traj", consecutively counting up. If any of these files
+   * cannot be loaded, the method returns null.
+   *
+   * @param trajName The path name in Choreo for this trajectory.
+   * @return The ArrayList of segments, in order, or null.
+   */
+  public static ArrayList<ChoreoTrajectory> getTrajectoryGroup(String trajName) {
+    var traj_dir = new File(Filesystem.getDeployDirectory(), "choreo");
```
Comment (calcmogul) [inline choreolib/src/main/java/com/choreo/lib/Choreo.java:54]: This code is pretty messy. Consider adding some blank lines to logically separate different pieces.

**Comment by calcmogul** [comment]
Actually, this whole vendordep has a lot of documentation issues, including in code outside this PR. Did we ever publish Doxygen/JavaDoc docs for ChoreoLib?

**Review by shueja — APPROVED** [review]

**Review by calcmogul — APPROVED** [review]

**Pull request merged**